torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -141,33 +186,51 @@ struct GraphFuser {
   }
 
   bool hasSupportedType(Node* node) {
-    return areTensorsOfSameShape(node->inputs()) &&
-           haveSupportedType(node->inputs()) &&
+    return haveSupportedShape(node->inputs()) &&
+           haveSupportedType(node->inputs(), /*allow_constant_numbers*/true) &&
            haveSupportedType(node->outputs());
   }
 
-  bool haveSupportedType(at::ArrayRef<Value*> list) {
+  bool haveSupportedType(at::ArrayRef<Value*> list, bool allow_constant_numbers=false) {
     for (Value *v : list) {
-      if (!hasSupportedType(v)) return false;
+      if (!hasSupportedType(v, allow_constant_numbers)) return false;
     }
     return true;
   }
 
+  bool haveSupportedShape(at::ArrayRef<Value*> list) {
+    std::vector<Value*> tensor_values;
+    tensor_values.reserve(list.size());
+    for (Value * v : list) {
+      if (v->type()->cast<TensorType>()) {
+        tensor_values.push_back(v);
+        continue;
+      }
+      // XXX: For now, we support fusing tensor-scalar ops if the scalars
+      // are constant. They are directly inlined into the kernel body
+      if (v->type()->isSubtypeOf(NumberType::get()) &&
+          v->node()->kind() != prim::Constant) {
+        return false;
+      }
+    }
+    if (tensor_values.size() == 0) {
+      return true;
+    }
+    return areTensorsOfSameShape(tensor_values);
+  }
+
 
   bool isFusable(Node * node) {
     if (node->owningBlock() != block) return false;
     if (node->kind() == prim::FusionGroup) return true;
     if (!isSimpleMap(node)) return false;
 
-    if (node->matches("aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor", /*const=*/attr::alpha)) {
-      std::vector<Value*> inputs {node->namedInput(attr::self), node->namedInput(attr::other)};
-      return areTensorsOfSameShape(inputs) && haveSupportedType(inputs);
-    } else if (node->matches("aten::lt(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::le(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::gt(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::ge(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::eq(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::ne(Tensor self, Tensor other) -> Tensor")) {
+    if (node->matches("aten::lt(Tensor self, Tensor other) -> Tensor") ||
```

Comment:
Actually, since we have some `matches` code here, you might want to put it in there instead of in `hasSupportedType`